src/App.tsx

```diff
@@ -1,24 +1,162 @@
-/* eslint-disable max-len */
+/* eslint-disable react-hooks/rules-of-hooks */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, {
+  useMemo,
+  useEffect,
+  useState,
+  useCallback,
+} from 'react';
 import { UserWarning } from './UserWarning';
+import { Todo } from './types/Todo';
+import { createTodo, deleteTodo, getTodos } from './api/todos';
+import { TodoList } from './components/TodoList';
+import { TodoForm } from './components/TodoForm';
+import { Error } from './components/Error';
+import { Filters } from './types/Filters';
+import { Filter } from './components/Filter';
 
-const USER_ID = 0;
+const USER_ID = 10892;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState(Filters.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodo, setLoadingTodo] = useState([0]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then((todosFromServer) => setTodos(todosFromServer))
+      .catch(() => setError('Unable to load todos'));
+  }, []);
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
+
+    if (error) {
+      timeout = setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [error]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const uncompletedTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const filteredTodos = useMemo(() => {
+    switch (filter) {
+      case Filters.COMPLETED:
+        return completedTodos;
+
+      case Filters.ACTIVE:
+        return uncompletedTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
+  const handleCloseError = () => {
+    setError(null);
+  };
+
+  const addTodo = useCallback(async (title: string) => {
+    try {
+      const newTodo = {
+        userId: USER_ID,
+        completed: false,
+        title,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos((prevTodos) => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  }, []);
+
+  const showTitleError = (erorMessage: string) => {
+    setError(erorMessage);
+  };
+
+  const removeTodo = useCallback(async (todoId: number) => {
+    try {
+      setLoadingTodo(prevTodoId => [...prevTodoId, todoId]);
+      await deleteTodo(todoId);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setTempTodo(null);
```

Comment:
Why do we need to set it to null? I don't think, we're changing it during remove 💭

Reply:
U are absolutely right! I just copied this part from addTodo and forgot remove it🤦🏽‍♀️